### PR TITLE
Extend processors and csv file sink

### PR DESCRIPTION
**Pull request by johanneshiry**
Partly solves #73 (introduces an `AssetInputProcessor` for all `AssetInput` entities)

when this one is merged, we still need processors for remaining `InputEntity`s (`GraphicInput`, `OperatorInput`, `AssetCharacteristicInput`-subclasses, `RandomLoadParameter` and `AssetTypeInput` - _when addressing this, it might be possible to extend the `AssetInputProcessor` to a full `InputEntityProcessor` instead of splitting up logic to new processors_) 

**Additionally to #73 when reviewing @ckittl please carefully check for #79**

### CODE REVIEW

**Comment by johanneshiry** [comment]
!test

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+import edu.ie3.test.common.GridTestData
+import edu.ie3.test.common.SystemParticipantTestData
+import edu.ie3.util.TimeTools
+import spock.lang.Specification
+
+import java.time.ZoneId
+import java.time.ZonedDateTime
+
+/**
+ * //ToDo: Class Description
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]: 😁

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+ *
+ * @version 0.1* @since 24.03.20
+ */
+class AssetInputProcessorTest extends Specification {
+
+
+    def "A ResultEntityProcessor should de-serialize a provided NodeInput correctly"() {
+        given:
+        TimeTools.initialize(ZoneId.of("UTC"), Locale.GERMANY, "yyyy-MM-dd HH:mm:ss")
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]: Didn't we agree on ISO strings?

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+                "uuid"         : "5dc88077-aeb6-4711-9142-db57292640b1",
+                "geoPosition"  : "{\"type\":\"Point\",\"coordinates\":[7.411111,51.492528],\"crs\":{\"type\":\"name\",\"properties\":{\"name\":\"EPSG:4326\"}}}",
+                "id"           : "node_a",
+                "operatesUntil": "2020-03-25 15:11:31",
+                "operatesFrom" : "2020-03-24 15:11:31",
+                "operator"     : "8f9682df-0744-4b58-a122-f0dc730f6510",
+                "slack"        : "true",
+                "subnet"       : "1",
+                "vTarget"      : "1.0",
+                "voltlvl"      : "Höchstspannung (380 kV)",
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]: Just for homogeneity: At other places I expect the voltage level to be of the form `"{Höchstspannung, 380 kV}"`. Additionally: Shouldn't this be de-serialized to
```
"voltlvl"      : "Höchstspannung"
"vRated"    : "380 kV"
```
?

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+ * @version 0.1* @since 24.03.20
+ */
+class AssetInputProcessorTest extends Specification {
+
+
+    def "A ResultEntityProcessor should de-serialize a provided NodeInput correctly"() {
+        given:
+        TimeTools.initialize(ZoneId.of("UTC"), Locale.GERMANY, "yyyy-MM-dd HH:mm:ss")
+        def sysPartResProcessor = new AssetInputProcessor(NodeInput)
+        def validResult = GridTestData.nodeA
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]:
```suggestion
        def validEntity = GridTestData.nodeA
```

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+ *
+ * @version 0.1* @since 24.03.20
+ */
+class AssetInputProcessorTest extends Specification {
+
+
+    def "A ResultEntityProcessor should de-serialize a provided NodeInput correctly"() {
+        given:
+        TimeTools.initialize(ZoneId.of("UTC"), Locale.GERMANY, "yyyy-MM-dd HH:mm:ss")
+        def sysPartResProcessor = new AssetInputProcessor(NodeInput)
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]: Too much of copy, paste. ;-) Please go through the test and adapt the other places as well.

```suggestion
        def assetInputProcessor = new AssetInputProcessor(NodeInput)
```

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        def validInput = modelInstance
+
+        when: "the entity is passed to the processor"
+        def processingResult = sysPartResProcessor.handleEntity(validInput)
+
+
+        then: "make sure that the result is as expected "
+        processingResult.present
+
+        println "["
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]: Do you still need this? If so, please put into a logger.

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+                "olmCharacteristic"  : "olm",
+                "operatesUntil"      : "2020-03-25 15:11:31",
+                "operatesFrom"       : "2020-03-24 15:11:31",
+                "operator"           : "8f9682df-0744-4b58-a122-f0dc730f6510",
+                "type"               : "3bed3eb3-9790-4874-89b5-a5434d408088"
+        ]
+
+    }
+
+    def "A ResultEntityProcessor should de-serialize a provided SystemParticipantInput correctly"() {
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]:
```suggestion
    def "A AssetInputProcessor should de-serialize a provided SystemParticipantInput correctly"() {
```

**Comment by ckittl** [comment]
> Partly solves #73 (introduces an `AssetInputProcessor` for all `AssetInput` entities)
> 
> when this one is merged, we still need processors for remaining `InputEntity`s (`GraphicInput`, `OperatorInput`, `AssetCharacteristicInput`-subclasses, `RandomLoadParameter` and `AssetTypeInput` - _when addressing this, it might be possible to extend the `AssetInputProcessor` to a full `InputEntityProcessor` instead of splitting up logic to new processors_)
> 
> **Additionally to #73 when reviewing @ckittl please carefully check for #79**

This PR closes #79. I checked thoroughly. The found mistake in complex topology actual brought a different topology then expected, but the mixed up nodes are from the same sub grid. Therefore, the tests for sub grid topology generation are correct and working. 👍

**Comment by ckittl** [comment]
!test

**Comment by ckittl** [comment]
!test

File: docs/uml/InputDatamodelConcept.puml [hunk 8]
```diff
@@ -8,11 +8,9 @@ abstract Class InputEntity
 InputEntity --|> UniqueEntity
 
 Interface Operable {
-+ operationInterval: Interval<ZonedDateTime>
-+ operatesFrom: ZonedDateTime
-+ operatesUntil: ZonedDateTime
 + inOperation: Boolean
-+ operator: OperatorInput
++ getOperator(): OperatorInput
```
[the remark this answers is not in the document]
Comment (johanneshiry) [reply docs/uml/InputDatamodelConcept.puml]: I know but I thought it would make sense for an interface - anyway I'll remove it

File: src/main/java/edu/ie3/datamodel/io/factory/input/AssetInputEntityFactory.java [hunk 9]
```diff
@@ -25,8 +25,8 @@
     extends EntityFactory<T, D> {
 
   private static final String UUID = "uuid";
-  private static final String OPERATES_FROM = "operatesfrom";
-  private static final String OPERATES_UNTIL = "operatesuntil";
+  public static final String OPERATES_FROM = "operatesfrom";
```
[the remark this answers is not in the document]
Comment (johanneshiry) [reply src/main/java/edu/ie3/datamodel/io/factory/input/AssetInputEntityFactory.java]: good catch - used them somewhere else during dev but looks like I only partly reversed corresponding changes

File: src/main/java/edu/ie3/datamodel/io/processor/EntityProcessor.java [hunk 10]
```diff
@@ -37,10 +43,20 @@
   public final Logger log = LogManager.getLogger(this.getClass());
   private final Class<? extends T> registeredClass;
   protected final String[] headerElements;
-  protected final LinkedHashMap<String, Method> fieldNameToMethod = new LinkedHashMap<>();
+  private final Map<String, Method> fieldNameToMethod;
+
+  private static final String OPERATION_TIME_FIELD_NAME = OperationTime.class.getSimpleName();
+  private static final String OPERATES_FROM = "operatesFrom";
+  private static final String OPERATES_UNTIL = "operatesUntil";
```
[the remark this answers is not in the document]
Comment (johanneshiry) [reply src/main/java/edu/ie3/datamodel/io/processor/EntityProcessor.java]: had exactly this before but finally reverted it again due to differences in upper/lower case - would like to keep it separated

File: src/main/java/edu/ie3/datamodel/io/processor/EntityProcessor.java [hunk 11]
```diff
@@ -37,10 +43,20 @@
...
+
+  private static final String VOLT_LVL_FIELD_NAME = "voltLvl";
+  private static final String VOLT_LVL = NodeInputFactory.VOLT_LVL;
+  private static final String V_RATED = NodeInputFactory.V_RATED;
+
+  private static final GeoJsonWriter geoJsonWriter = new GeoJsonWriter();
 
   /** Field name of {@link UniqueEntity} uuid */
-  private static final String uuidString = "uuid";
+  private static final String UUID_FIELD_NAME = "uuid";
```
[the remark this answers is not in the document]
Comment (johanneshiry) [reply src/main/java/edu/ie3/datamodel/io/processor/EntityProcessor.java]: because it directly refers to the field/attribute name in a entity we are processing, while in contrast all other values without the `_FIELD_NAME` suffix are "derived fields" in the sense that they are not fields in our main object instance we are processing but wrapped in e.g. `OperationTime` or `VoltLevel` instances - hence `field` in this case is the real field in the entity

File: src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy (hunk 3 above)
Comment (johanneshiry) [reply src/test/groovy/edu/ie3/datamodel/io/processor/input/AssetInputProcessorTest.groovy]: No. ;-)

we should have a short talk about this then.

tl;dr-version:

1. voltLvl is the `id` String instance you placed in `GermanVoltageLevelUtils`. After our short discussion yesterday this is (from a serialized perspective) equal to one column in a resulting .csv-file or database field named `voltLvL`. We agreed yesterday to keep this structure to not introduce new columsn/fields. If you think it makes sense to change this again let me know. As discussed several times previously, I'm not a big fan of putting json strings inside column/row based data structures as this can become very hard to read quickly
2. Your second commend is correct, IF we want to introduce two new fields in our datastructs. If you think this makes sense, we should discuss this. 
3. Where du you expect `{Höchstspannung, 380 kV}` in other places? As this is neither standard json encoding nor something else I would highly recommend changing this to a clearer (maybe more suitable) structure.

**Comment by johanneshiry** [comment]
!test
